core/src/main/java/io/substrait/expression/ExpressionCreator.java

```diff
@@ -102,17 +114,59 @@ public static Expression.TimestampLiteral timestamp(
             .withNano((int) TimeUnit.MICROSECONDS.toNanos(micros)));
   }
 
+  /**
+   * @deprecated TimestampTZ is deprecated in favor of PrecisionTimestampTZ
+   */
+  @Deprecated
   public static Expression.TimestampTZLiteral timestampTZ(boolean nullable, long value) {
     return Expression.TimestampTZLiteral.builder().nullable(nullable).value(value).build();
   }
 
+  /**
+   * @deprecated TimestampTZ is deprecated in favor of PrecisionTimestampTZ
+   */
+  @Deprecated
   public static Expression.TimestampTZLiteral timestampTZ(boolean nullable, Instant value) {
     var epochMicro =
         TimeUnit.SECONDS.toMicros(value.getEpochSecond())
             + TimeUnit.NANOSECONDS.toMicros(value.getNano());
     return timestampTZ(nullable, epochMicro);
   }
 
+  public static Expression.PrecisionTimestampLiteral precisionTimestamp(
+      boolean nullable, long value, int precision) {
+    return Expression.PrecisionTimestampLiteral.builder()
+        .nullable(nullable)
+        .value(value)
+        .precision(precision)
+        .build();
+  }
+
+  public static Expression.PrecisionTimestampLiteral precisionTimestamp(
+      boolean nullable, LocalDateTime value) {
+    var epochMicro =
+        TimeUnit.SECONDS.toMicros(value.toEpochSecond(ZoneOffset.UTC))
+            + TimeUnit.NANOSECONDS.toMicros(value.toLocalTime().getNano());
+    return precisionTimestamp(nullable, epochMicro, 6);
```

Comment:
The docs for [LocalDateTime](https://docs.oracle.com/javase/8/docs/api/java/time/LocalDateTime.html) indicate
> Time is represented to nanosecond precision.

Should we be using nanoseconds (9) here instead of microseconds (6)?

Reply:
I have no real preference here - just kept it like this so that it aligns as closely as possible to the existing timestamp creators.

Given the LocalDate (and also Instant) contain more data than just a int64, (they have seconds and then the subseconds separately), I guess there's some dates that cannot be represented as only nanonseconds but can in microseconds. Quick search suggests that's about +- 292 years. So while we can switch this to nanoseconds, it could overflow for some inputs (if those inputs are like year 3000).

Frankly, I think I'd rather keep this as is, to make switching from the earlier timestamp behavior as easy as possible. If people want more control, they can always use the other creator func that lets them decide the precision. Alternatively, we could even consider not adding this second constructor. Thoughts?

Reply:
> Alternatively, we could even consider not adding this second constructor. 

Let's leave the new constructors out for now. It sounds like we may need to consider their behaviours more carefully now that we're not eliding to microseconds by default, but we don't need to wait to figure that out before merging the rest of this in.

Reply:
Done! https://github.com/substrait-io/substrait-java/pull/283/commits/f255e6c1293488740b8b5100e4bb0a10d572f554